### PR TITLE
fix silenced JSON parsing errors

### PR DESCRIPTION
**Pull request by khoomeister**

### CODE REVIEW

**Comment by jkrems** [comment]
:heart: for the increased coverage. :shipit:

File: test/json.test.coffee [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      result = safeParseJSON '{"test": 1}', res
+      assert.deepEqual {GOFER_PARSED_RESPONSE: true}, res
+      assert.deepEqual
+        parseError: null
+        body: {test: 1}
+      , result
+
+    it "will return an empty object if response doesn't exist", ->
+      result = safeParseJSON '{"test": 1}'
+      assert.deepEqual {}, result
```
Comment (johan) [inline test/json.test.coffee]: Might I suggest combining these two lines to the single-line use case Assertive was designed for?

```coffeescript
      assert.deepEqual {}, safeParseJSON '{"test": 1}'
```

File: src/json.coffee [hunk 2]
```diff
@@ -33,17 +33,18 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 @safeParseJSON = (str, res) ->
+  return {} unless res?
   res[GOFER_PARSED] = true
 
-  data = error: null
+  data = parseError: null
   try
     data.body =
-      if 'string' == typeof str
+      if 'string' == typeof str and str != ''
```
Comment (khoomeister) [inline src/json.coffee]: @jkrems for 304s, an empty body is returned, so we shouldn't be parsing it regardless of what parseJSON, etc. is saying... does that rule make sense?

File: test/json.test.coffee (hunk 1 above)
Comment (khoomeister) [inline test/json.test.coffee]: Sounds good :smiley:

File: src/json.coffee (hunk 2 above)
Comment (jkrems) [inline src/json.coffee]: If a service sends an empty string (e.g. as a 200) and a user explicitly said they expected JSON, we should fail. We already check for empty strings in `isJsonResponse`. If a consumer is opting out of the auto-detection and forces parsing, they should know what they are doing.

Comment (khoomeister) [inline src/json.coffee]: Since we also have the response object, could we check for the 304 status code then?

Comment (jkrems) [inline src/json.coffee]: If we reach this point with a 304 the following *should* be true:

* The user explicitly forced parsing
* The user explicitly allowed status codes >= 300

I don't think this function should do any more than stupidly parse json. Anything else will only hide bugs in application logic or configuration.

Comment (khoomeister) [inline src/json.coffee]: OK makes sense - I've removed the empty string check.

**Comment by jkrems** [comment]
```
1) Basic Integration Test combined timeouts does not pass an error when timeout and completion timeouts are not exceeded:
     Uncaught Error: ETIMEDOUT
    at [object Object].<anonymous> (/home/travis/build/groupon/gofer/node_modules/request/request.js:713:15)
    at Timer.listOnTimeout (timers.js:88:15)
```

Looks like one of the timeout tests is flaky?

**Comment by jkrems** [comment]
:shipit: when green!

**Comment by khoomeister** [comment]
2.3.5

**Pull request merged**